% !TeX root = ../main.tex
\documentclass[../main.tex]{subfile}
\begin{document}\pagebreak\section*{Abstract}
    The utilisation of the canonical stop codons (TAA, TGA, TAG) is dependent on various factors, including the genomic GC content. Its been shown in bacteria that relative usage of TAA and TGA is correlated with the GC content of the genome (decreasing with GC in the case of TAA and increasing for TGA), however TAG is not correlated with GC content and is utilised less. Its been hypothesis that usage of stop codons co-evolved with the genomic GC content and the expression ratio of the two codon recognising release factors in bacteria (RF1 and RF2), which bind to either TGA or TAG (as well as TAA) -- RF1:RF2 hypothesis. Archaea and eukaryotes, however utilise a single release factor to bind all three codons.

    The usage of these codons was explored in bacteria and archaea, using a phylogenetically controlled dataset, as well as in the human genome. It was found that in all three domains, the usage of stop codons followed the aforementioned usage pattern, showing up again downstream of the stop coding of protein coding sequences in archaea and bacteria, where RF1 and RF2 shouldn't influence codon usage. The relative usage of stop codons in tRNA sequences was explored for bacteria, it appeared as though TAG was more highly utilised that TGA or TAA, but given that all three codons were independent of GC, it suggested that tRNA is likely selected to maintain stability.

    Given that the TAG was underutilised in all other cases, it strongly suggests that RF1:RF2 hypothesis does not explain the behaviour of stop codon usage in bacteria.

\end{document}
