\documentclass[
    emulatestandardclasses,
    parskip=half,
]{scrartcl}
\usepackage{scrhack}

\usepackage[a4paper,
    top=1.25in,
    left=.75in,
    bottom=1.5in,
    right=.75in
]{geometry}
\usepackage{subfiles}

\usepackage[british]{babel}
\usepackage[utf8]{inputenc}
\usepackage{fontenc}

\usepackage{csquotes,graphicx,caption,subcaption,float,wrapfig}
\usepackage{booktabs,multirow,makecell,longtable,threeparttable,tabularx}
\usepackage{pdflscape}
\usepackage{pdfpages}
\usepackage[toc, page]{appendix}

\usepackage[
	separate-uncertainty   = true,
	multi-part-units       = single,
	range-units            = single,
	list-units             = single,
	table-number-alignment = center,
	table-text-alignment   = center
]{siunitx}
\usepackage{gensymb,mathtools,amssymb,xfrac}
\usepackage[italicdiff]{physics}
\usepackage{upgreek,chemmacros,chemfig}


\usepackage[hidelinks]{hyperref}
\usepackage[nameinlink,noabbrev]{cleveref}
\usepackage[style=authoryear-ibid]{biblatex}
\usepackage{todonotes}

\let\mc\multicolumn
\let\mr\multirow

% \maxdeadcycles=1000

\DeclareSIUnit\G{G}
\chemsetup{ modules = all }
\chemsetup[reactions]{
	tag-open = (,
	tag-close =)
}

% \setlength{\parskip}{1em}

\addbibresource{biblio.bib}

\newcommand{\textapprox}{\raisebox{0.5ex}{\texttildelow}}
\newcommand{\RNum}[1]{\uppercase\expandafter{\romannumeral #1\relax}}

\begin{document}
    \title{\textit{(Working Title)} UAG Codon Paradox and the RF1:RF2 Hypothesis}
    \date{\today}
    \author{Sulaiman Sulaiman}
    \maketitle

    \begin{abstract}
        {\textbf{Abstract}\vspace{1em}}

        Abstract goes here
    \end{abstract}

    \section{Introduction}
        The termination of mRNA translation into proteins is classically controlled by a set of three stop codons: UAA, UAG, and UGA. Recognition of these codons triggers a choreographed sequence of events which result in the dismantling of the ribosome releasing the synthesised nascent peptide chain and the mRNA. Critical to this release process are release factors proteins - these proteins mimic the structure and function of tRNA \autocite{Nakamura2003}, recognising \& binding to the stop sequence and triggering the dismantling process.

        In bacteria, three release factors (RF) are present of which RF1 (which recognises UAA and UAG) as well as RF2 (which recognises UAA and UGA) perform the recognition, with RF3 acting to release RF1/2 from the ribosome. Eukaryotes and Archaea, contain a single release factor which recognises the stop codons, called eRF1 and aRF1 respectively, these proteins share a various features \autocite{Alkalaeva2009}, so much so that aRF1 is able to replace the function of eRF1 \autocite{Dontsova2000}, both release factors are capable of recognising all three termination codons, unlike their bacterial counterparts.

        \begin{wrapfigure}{r}{7.5cm}
            \centering
            \includegraphics[
                clip, trim={0 5.5cm 0 0}
            ]{"../plot/references/usage-vs-gc.jpg"}
            \caption{Distribution of stop codon usage against GC of bacteria from \mancite\cite{Korkmaz2014}}
            \label{fig:intro/codon-gc}
        \end{wrapfigure}

        The usage of these stop codons is under various pressures, including reducing readthrough rates, stopping out-of-frame reading and GC pressure on the genome. It has been observed that as the GC content of a bacterial genomes increases, the prevalence of UAA stop codons decreases while UGA increases, these codons follow what would be expected given their different GC content. Meanwhile, UAG breaks this trend, with its usage remaining low (20\%) and independent of the GC content of the genome, countering what otherwise be expected especially given how the other codons follow the pattern, \cref{fig:intro/codon-gc}. One theory postulates that usage of UGA and UAG co-evolved with the relative expression of RF1 and RF2, \mancite\autocite{Korkmaz2014}, (henceforth, referred to as the RF1:RF2 hypothesis). Whether, the RF1:RF2 hypothesis properly explains is unresolved. Recently the same trends have been seen in downstream addition stop codons, \autocite{Ho2019}, in bacteria. This might possibly reflect selection for termination if these are either fail-safe stop codons to trap ribosomes that have read through the primary stop. Similarly, of frame stop codons may function to catch frameshifted ribosomes.

        One could postulate that effects seen downstream may still be within the purview of RF1/2 through some other mechanism. As such, to properly isolate the influence of these proteins and to properly identify whether RF1:RF2 hypothesis holds true or if another mechanism is underlying the patterns seen, this project aims to investigating the usage of the stop codons in Eukaryotes and Archaea, given that they utilise a single release factor for the recognition of all termination codons, at the terminus of protein coding sequences, as well as downstream additional codons. Additionally usage of UAG and UGA in non-coding RNA (such as tRNA and rRNA) will be investigated, as it would be expected that any usage of the stop codons should only be correlated with GC, if the RF1:RF2 hypothesis is correct, as these do not participate in translation.

    \section{Methods}
        The bacterial and archaeal datasets used were previously created by Laurence Hurst - these dataset was created by filtering out all but the largest genome of each genus to ensure that only one representative genome was left per genus.

        To ensure that the protein coding genes which are used for the analysis are indeed protein coding genes they were checked against a couple basic criteria. This was run on features marked as CDS (Coding Sequence), in part as these should mostly be expected to pass. First, the beginning codon and end codon were checked against the translation table associated with each gene, this was to ensure they were valid start and stop codons respectively. Then to ensure that genes spanned a proper reading frames the length of protein coding genes was check for divisibility by 3. Finally, internal stop codons were checked for as these would indicate the gene was likely non-functional and thus not likely to be under the same influences.

        \begin{table}[H]
            \centering
            \caption{Number of CDS annotated features which failed in each category}
            \begin{tabular}{ r c c c c }
                \toprule
                              & \mc{2}{c}{Orignal}     & \mc{2}{c}{Without pseudogenes} \\
                                \cmidrule(rl){2-3}       \cmidrule(rl){4-5}
                Category      & {Bacteria} & {Archaea} & {Bacteria} & {Archaea} \\
                \midrule
                Start         & 687        & 251         & 106        & 2        \\
                Stop          & 843        & 163         & 40         & 0        \\
                Internal Stop & 1745       & 331         & 503        & 74        \\
                Length        & 736        & 158         & 93         & 0        \\
                \bottomrule
            \end{tabular}
        \end{table}

        Account for pseudogenes, by removing them from the list, was followed by a substantial reducing in the number of CDS features which failed the check

        By accounting for pseudogenes (any of the features with the "pseudo" annotation) and removing them from the list of genes which failed, a large drop was seen showing that they account for the vast majority of these genes, accounting for \SI{0}{\percent} of all failures. A record of these features was kept to remove them from any further analysis.

        All the scripts used can be seen on \href{https://github.com/TheKingOfAtlantis/BScProject}{Github}
    \section{Results}
        \subsection{Is pattern maintained when controlled for phylogeny?}
        \subsection{What pattern do out-of-frame stop codons follow?}
        \subsection{Do these patterns hold for ncRNA (tRNA + rRNA)?
                    As stop codons not used (not read by ribosome)}
        \subsection{Does the patterns seen extend to Eukaryotes and Archaea?
                    As they utilise a single termination factor}
    \section{Discussion}

    \section{Conclusions}

	\newpage \phantomsection
	\addcontentsline{toc}{chapter}{Bibliography}
	\printbibliography
	\appendix\appendixpage
\end{document}
